presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -46,7 +47,16 @@
     @Override
     public Optional<PlanNode> apply(WindowNode parent, Captures captures, Context context)
     {
-        WindowNode child = captures.get(CHILD);
+        // Pulling the descendant WindowNode above projects is done as a part of this rule, as opposed in a
+        // separate rule, because that pullup is not useful on its own, and could be undone by other rules.
+        // For example, a rule could insert a project-off node between adjacent WindowNodes that use different
+        // input symbols.
```

Comment:
@kokosing 
> WindowNode sorting on its own has not much sense.
It DOES make sense. The purpose of sorting is not only to make it eligible for merge but also to make windows with the same `partition by` clause adjacent, therefore we need less repartitionings to execute them, so even without `MergeAdjacentWindows`, `SwapAdjacentWindows` makes sense and speeds up some queries.
Imagine WFs as follows:
```
sum over (partition by A order by C)
avg over (partition by B)
sum over (partition by A order by D)
```
if you reorder them (`SwapAdjacentWindows` currently does it) into:
```
avg over (partition by B)
sum over (partition by A order by D)
sum over (partition by A order by C)
```
you need only 2 repartitionings instead of 3, as you'd need in the former plan.
Therefore, even though you still cannot merge any windows in such an example, the plan is better after the reordering.

I agree with @alandpost that creating a separate block of iterative rules kills benefits of iterative optimizer, preventing the rule of making window nodes mergeable in some cases which we didn't come up with while writing the rule originally.
We should keep the rules self-content and ensure they preserve the invariant of correct plan for every input the rule accepts, not more than that.
Rest should be explored by iterative-element of the optimizer.

Then it's easy to reason about the whole optimizer being correct without testing gigantic search space for all possible plans.